src/tools/cephfs/cephfs_shell.py

```diff
@@ -0,0 +1,396 @@
+    print('File size: {}'.format(file_size))
+    progress = 0
+    for chunk_start,chunk_size in get_chunks(file_size):
+        file_chunk = cephfs.read(file_,chunk_start, chunk_size)
+        progress += len(file_chunk)
+        if fd > 0:
+            fd.write(file_chunk)
+            print('{0} of {1} bytes read ({2}%)'.format(progress, file_size, int(progress / file_size * 100))) 
+        else :
+            print(file_chunk)   
+    cephfs.close(file_)
+    if fd > 0: 
+        fd.close()   
+
+class CephfsShell(cmd.Cmd):
+
+    def __init__(self):
+        super().__init__()
+        self.set_prompt()
+        self.intro = 'Ceph File System Shell'
+        self.interactive = False
+        self.parser = argparse.ArgumentParser(description='CephFS Shell.')
+        
+    def default(self, line):
+        print("Unrecognized command:", line)
+
+    def set_prompt(self):
+        self.prompt = '\033[01;33mCephFS >>>\033[00m ' 
+   
+    def create_argparser(self, command):
+        try:
+            argparse_args = getattr(self, "argparse_" + command)
+        except AttributeError:
+            return None
+        doc_lines = getattr(self, "do_" + command).__doc__.expandtabs().splitlines()
+        if '' in doc_lines:
+            blank_idx = doc_lines.index('')
+            usage = doc_lines[:blank_idx]
+            description = doc_lines[blank_idx + 1:]
+        else:
+            usage = doc_lines
+            description = []
+        parser = argparse.ArgumentParser(
+            prog=command,
+            usage='\n'.join(usage),
+            description='\n'.join(description),
+            formatter_class=argparse.ArgumentDefaultsHelpFormatter
+        )
+        for args, kwargs in argparse_args:
+            parser.add_argument(*args, **kwargs)
+        return parser
+
+    
+    def onecmd(self, line):
+        """Global error catcher"""
+        try:
+            res = cmd.Cmd.onecmd(self, line)
+            if self.interactive: 
+                self.set_prompt()
+            return res
+        except ConnectionError as e:
+            print("***", e)
+        except KeyboardInterrupt:
+            print("Command aborted")
+        except Exception as e:
+            print(e)
+            traceback.print_exc(file=sys.stdout)
+
+    def do_mkdir(self,arglist):
+        """Create directory.
+           Usage: mkdir [-m MODE] <dir_name> """    
+        try:
+            self.parser = argparse.ArgumentParser(description='Create Directory.')
+            self.parser.add_argument('dir_name', type = str, help = ' Name of new_directory.')
+            self.parser.add_argument('-m', '--mode', action='store', help='Sets the access mode for the new directory.',type=str)
+            args = self.parser.parse_args(shlex.split(arglist))
+            path = bytes('/' + args.dir_name, encoding="ascii")
+            if args.mode :
+                permission = int(args.mode,8)
+            else :
+                permission = 0o777
```

Comment:
This is fine but you should provide a way to set the `umask(3)` of the shell session. Unfortunately we haven't written a binding yet for setting the umask in libcephfs.c. I have created an issue:

http://tracker.ceph.com/issues/24430

I suggest spending a little time familiarizing yourself with the umask command. We can discuss how to implement it tomorrow.